mapgen/iwyu-mapgen-windowsh.py

```diff
@@ -0,0 +1,213 @@
+from collections.abc import Generator
+
+IF_RE = re.compile(r"#\s*if")
+ENDIF_RE = re.compile(r"#\s*endif")
+LEAN_AND_MEAN1_RE = re.compile(r"#\s*ifndef\s+WIN32_LEAN_AND_MEAN")
+LEAN_AND_MEAN2_RE = re.compile(r"#\s*if\s+!defined\s*\(WIN32_LEAN_AND_MEAN\)")
+
+INCLUDE_RE = re.compile(r"#\s*include\s+<(.+)>")
+
+
+# Standard C Library headers as of C23
+STDLIB_EXCLUSIONS = [
+    "assert.h",
+    "complex.h",
+    "ctype.h",
+    "errno.h",
+    "fenv.h",
+    "float.h",
+    "inttypes.h",
+    "iso646.h",
+    "limits.h",
+    "locale.h",
+    "math.h",
+    "setjmp.h",
+    "signal.h",
+    "stdalign.h",
+    "stdarg.h",
+    "stdatomic.h",
+    "stdbit.h",
+    "stdbool.h",
+    "stdckdint.h",
+    "stddef.h",
+    "stdint.h",
+    "stdio.h",
+    "stdlib.h",
+    "stdmchar.h",
+    "stdnoreturn.h",
+    "string.h",
+    "tgmath.h",
+    "threads.h",
+    "time.h",
+    "uchar.h",
+    "wchar.h",
+    "wctype.h",
+]
+
+
+accum_includes: set[str] = set()
+
+
+def parse_include_names(headerpath: Path) -> Generator[str]:
+    """
+    Parse the header file at headerpath and return all include names.
+
+    This excludes headers that are guarded by the WIN32_LEAN_AND_MEAN macro,
+    it's generally recommended that users should include those manually, so we
+    don't want to map them to <Windows.h>.
+    """
+
+    if_count = 0
+    lean_mean_idx: int | None = None
+
+    with headerpath.open("r", encoding="utf-8") as fobj:
+        for line in fobj.readlines():
+            if IF_RE.search(line) is not None:
+                # Check if we entered a WIN32_LEAN_AND_MEAN block
+                if (
+                    LEAN_AND_MEAN1_RE.search(line) is not None
+                    or LEAN_AND_MEAN2_RE.search(line) is not None
+                ):
+                    lean_mean_idx = if_count
+
+                if_count += 1
+                continue
+
+            if ENDIF_RE.search(line) is not None:
+                if_count -= 1
+
+                # Check if we escaped a WIN32_LEAN_AND_MEAN block
+                if if_count == lean_mean_idx:
+                    lean_mean_idx = None
+
+                continue
+
+            if lean_mean_idx is None:
+                include_match = INCLUDE_RE.search(line)
+                if (
+                    include_match is not None
+                    and include_match.group(1) not in STDLIB_EXCLUSIONS
+                ):
+                    yield include_match.group(1)
+
+
+def find_file_in(include_path: Path, filename: str) -> Path | None:
+    found = list(include_path.rglob(filename))
+    if found:
+        return found[0]
+
+    return None
+
+
+def fill_descendant_includes_of(include_root: Path, header_name: str) -> None:
+    """Recursively find all includes of `header_name`.
+
+    Found includes are added into the `accum_includes` set.
+    """
+    file = find_file_in(include_root, header_name)
+    if file is None:
+        return
+    accum_includes.add(header_name)
```

Comment:
I'm not too happy about the mutable global. It would be nicer to pass the include set in as a mutable parameter.